tree/ntuple/v7/src/RNTupleModel.cxx

```diff
@@ -318,6 +322,43 @@ void ROOT::Experimental::RNTupleModel::AddField(std::unique_ptr<RFieldBase> fiel
    fFieldZero->Attach(std::move(field));
 }
 
+void ROOT::Experimental::RNTupleModel::AddSubfield(std::string_view qualifiedFieldName, REntry &entry,
+                                                   bool initializeValue) const
+{
+   auto field = FindField(qualifiedFieldName);
+   if (initializeValue)
+      entry.AddValue(field->CreateValue());
+   else
+      entry.AddValue(field->BindValue(nullptr));
+}
+
+void ROOT::Experimental::RNTupleModel::RegisterSubfield(std::string_view qualifiedFieldName)
+{
+   if (qualifiedFieldName.empty())
+      throw RException(R__FAIL("no field name provided"));
+
+   if (fRegisteredSubfields.find(std::string(qualifiedFieldName)) != fRegisteredSubfields.end())
+      throw RException(R__FAIL("subfield already registered"));
+
+   EnsureNotFrozen();
+
+   auto *field = FindField(qualifiedFieldName);
+   if (!field) {
+      throw RException(R__FAIL("could not find subfield \"" + std::string(qualifiedFieldName) + "\" in model"));
+   }
+
+   auto parent = field->GetParent();
+   while (parent && !parent->GetFieldName().empty()) {
+      if (parent->GetStructure() == ENTupleStructure::kCollection) {
+         throw RException(R__FAIL("registering a subfield in a collection is not supported"));
+      }
+      parent = parent->GetParent();
+   }
+
+   fRegisteredSubfields.emplace(qualifiedFieldName);
+   AddSubfield(qualifiedFieldName, *fDefaultEntry);
```

Comment:
Perhaps swap the two lines. If we ever throw in `AddSubfield`, we can avoid this way an extra step to clear the name.